src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/FullProjectSnapshotHandleJsonConverter.cs

```diff
@@ -26,30 +26,65 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
+                        break;
+                    case nameof(FullProjectSnapshotHandle.FilePath):
+                        if (reader.Read())
+                        {
+                            filePath = (string)reader.Value;
+                        }
+                        break;
+                    case nameof(FullProjectSnapshotHandle.Configuration):
+                        if (reader.Read())
+                        {
+                            configuration = RazorConfigurationJsonConverter.Instance.ReadJson(reader, objectType, existingValue, serializer) as RazorConfiguration;
+                        }
+                        break;
+                    case nameof(FullProjectSnapshotHandle.RootNamespace):
+                        if (reader.Read())
+                        {
+                            rootNamespace = (string)reader.Value;
+                        }
+                        break;
+                    case nameof(FullProjectSnapshotHandle.ProjectWorkspaceState):
+                        if (reader.Read())
+                        {
+                            projectWorkspaceState = serializer.Deserialize<ProjectWorkspaceState>(reader);
+                        }
+                        break;
+                    case nameof(FullProjectSnapshotHandle.Documents):
+                        if (reader.Read())
+                        {
+                            documents = serializer.Deserialize<DocumentSnapshotHandle[]>(reader);
+                        }
+                        break;
+                }
+            });
 
             // We need to add a serialization format to the project response to indicate that this version of the code is compatible with what's being serialized.
             // This scenario typically happens when a user has an incompatible serialized project snapshot but is using the latest Razor bits.
 
-            if (!obj.TryGetValue(SerializationFormatPropertyName, out var serializationFormatToken))
-            {
-                // Pre-serialization format release.
-                return null;
-            }
-
-            var serializationFormat = serializationFormatToken.Value<string>();
-            if (serializationFormat != ProjectSerializationFormat.Version)
+            if (string.IsNullOrEmpty(serializationFormat) || serializationFormat != ProjectSerializationFormat.Version)
             {
                 // Unknown serialization format.
                 return null;
             }
 
-            var filePath = obj[nameof(FullProjectSnapshotHandle.FilePath)].Value<string>();
-            var configuration = obj[nameof(FullProjectSnapshotHandle.Configuration)].ToObject<RazorConfiguration>(serializer);
-            var rootNamespace = obj[nameof(FullProjectSnapshotHandle.RootNamespace)].ToObject<string>(serializer);
-            var projectWorkspaceState = obj[nameof(FullProjectSnapshotHandle.ProjectWorkspaceState)].ToObject<ProjectWorkspaceState>(serializer);
```

Comment:
This contains all our project Tag Helpers ~228 `TagHelperDescriptors`. 

The JObject we created in https://github.com/dotnet/aspnetcore-tooling/pull/2239/files#diff-852597f64a9a7cc7947eed53801b201aL29 added all of the `TagHelperDescriptors` alongside the rest of the properties here. This would explain the huge perf hit of this one converter. This will be beneficial for memory as well. 😄